### DISCUSSION TITLE
Comparison against yagop/node-telegram-bot-api and other libraries

### DISCUSSION THREAD

**Discussion by ApayRus** [telegraf/telegraf#386] [General]
Hi! 
I have completed one project with [yagop/node-telegram-bot-api](https://github.com/yagop/node-telegram-bot-api). Unfortunately, your project appears in google far away, and in youtube there is many screencasts with yagop library. And I noticed your project too far, but I liked it because of laconic code.  

Please, can you explain main differences of your project from yagop's ? 
Sorry for issue, but I didn't find any other place for feedback to send this question.

**Comment by dotcypress** [comment]
Hi @Aparus 

Key differences:
- Full BotAPI support
- Lightweight deps http://npm.anvaka.com/#/view/2d/node-telegram-bot-api vs http://npm.anvaka.com/#/view/2d/telegraf
- [Send request in response to update](https://core.telegram.org/bots/faq#how-can-i-make-requests-in-response-to-updates)
- First class middleware support(killer feature)
- Tons of examples, [scaffolding tools](https://github.com/telegraf/create-bot), [zero config bot runner](https://github.com/telegraf/micro-bot)

**Comment by TiagoDanin** [comment]
...
- Error handling
- Faster
Telegraf is the best ❤️

**Comment by ApayRus** [comment]
@dotcypress , I think you should add this info into the main page , it's really useful for newbies who are seeking tg-framework at first time in their life.

**Comment by anatoly314** [comment]
Is there any significant difference now, two years later? It looks like [node-telegram-bot-api](https://github.com/yagop/node-telegram-bot-api) more popular. Don't want to start the holywar, just want to know which one to choose for my needs.

**Comment by KnorpelSenf** [comment]
First, I should say that I have used a number of different frameworks in production, but not node-telegram-bot-api yet (only for playing around).

That said, I believe that the points outlined above are still valid today. We're not far away from releasing 4.0 which ships with Telegram Bot API 5.0 support that was released less than a month ago. node-telegram-bot-api's latest commit was six weeks before that ([ref](https://github.com/yagop/node-telegram-bot-api/commit/fc24d0dc4d45fb3eabe97f43325d662f18bf64c0)).

`telegraf` is still using much fewer dependencies. This might not be the issue for you, but especially in serverless environments it is important to keep startup times low for cold starts, and [dependencies are the top three important things to keep in mind](https://medium.com/@duhroach/improving-cloud-function-cold-start-time-2eb6f5700f6) there.

It is a dealbreaker for me not to have TypeScript support, which node-telegram-bot-api does not provide. While having had mediocre TypeScript support in the past, `telegraf` is now close to being fully typed with advanced inference systems that have access to the full Telegram API, hence giving you every single property in autocomplete right at your fingertips. This is also coming in 4.0, the code is mostly written for that.

While a bot is a complete middleware system in `telegraf`, in node-telegram-bot-api it is a plain event emitter. For small projects you probably won't notice the difference in convenience, but if you are planning ahead, I have my doubts about node-telegram-bot-api being flexible enough to fit your needs. The existence of https://github.com/idchlife/node-telegram-bot-api-middleware proves this IMO.

`telegraf` is used less, when comparing npm downloads per week. I cannot say much about the ecosystem around the libraries. While I didn't really miss anything yet for `telegraf`, it might well be that node-telegram-bot-api has an impressive ecosystem that I did not find by looking at their GitHub page.

`telegraf` has worse docs. It is pretty simple to get up and running, there's examples, type annotations, and an active community, but honestly: I learned more about `telegraf` from looking into the implementation than having a nice guide that helps me fix bugs in corner cases. We're working on that, but of course you shouldn't base that decision on promises.

**Reply by wojpawlik** [reply]
`node-telegram-bot-api` has _some_ TypeScript support: https://www.npmjs.com/@types/node-telegram-bot-api

**Reply by wojpawlik** [reply]
I'd say Telegraf 4.0.2's docs (with search, accurate types and links to source) are better than ntba 0.51.0's.

**Comment by wojpawlik** [comment]
> Ntba is just the PHP of bot API libraries

`telegraf` is [lightweight](https://packagephobia.com/result?p=telegraf,node-telegram-bot-api), actively maintained, and has a number of types that have one job, and do it [relatively](https://github.com/telegraf/telegraf/issues/1203#issuecomment-736509064) well. By contrast, `node-telegram-bot-api`

0. Misses `Markup`, `Router`, scenes, and `session`.
1. Misses `Composer` and `Middleware`. It only has the equivalent of `Telegraf::hears` and `Telegraf::on`. As a result, I ended up inventing [a hack](https://bitbucket.org/GingerPlusPlus/rextester-bot/src/cddd0eae01d12d04c1dc4601e7a525287c7fbe87/main.ls#lines-103) to achieve something as simple as `bot.drop(ctx => ctx.message.forward_date !== undefined)`.
2. Misses `Context`. As a result, one has to import `node-telegram-bot-api`'s `TelegramBot` instance (which performs the job of `Telegram`) into every file with a handler, and pass `chatId` to every call, inconvenient!
3. Violates single-responsibility principle: `TelegramBot` also performs the functions of `Telegraf`, which leads to [circular imports](https://bitbucket.org/search?q=repo%3Arextester-bot%20cycle).
4. Is [heavy](https://packagephobia.com/result?p=telegraf,node-telegram-bot-api), despite lacking functionality and primitive design.
5. Depends on deprecated `request` and `request-promise`.
6. Is barely maintained, with only 16 commits on default branch in 2020.
   ```
   ~/JS/telegraf (develop $=)
   $ git log --after="2020-01-01 00:00" --before="2020-12-31 23:59" --oneline | wc -l
   175
   ```

Truly, there's no reason to choose `node-telegram-bot-api` over `telegraf`.

Fixing all of these issues would require a full rewrite.
So, why don't you deprecate `node-telegram-bot-api` in favor of `telegraf`, @kamikazechaser?

<hr>

Funnily enough, npm has several less popular bot API modules with the _exact same_ problems.

**Reply by kamikazechaser** [reply]
> Fixing all of these issues would require a full rewrite.

Yup, I agree.  We discussed with a few users on the roadmap and a rewrite is absolutely necessary. Telegraf wins almost everywhere except that it has a higher barrier to entry for newbie bot developers. `ntba` is much easier to pick up.

Deprecating it is not a solo decision that I can make. Also there are lots of tutorial out there using `ntba`.

**Reply by wojpawlik** [reply]
> We discussed [...] and a rewrite is absolutely necessary.

Link to the discussion? Do you have enough manpower for a rewrite? How would you achieve SRP?

> Telegraf [...] has a higher barrier to entry for newbie bot developers.

Sadly, I agree. I think it's due to having more classes (`Telegraf`, `Telegram`, `Context`), necessary for SRP.

**Reply by kamikazechaser** [reply]
> Link to the discussion?

Its on the official group, couple of weeks back. I also talked to Mugo regarding this last year. Regarding the rewrite, SRP, I'll have to sit down with past maintainers and discuss this.

**Reply by KnorpelSenf** [reply]
> Telegraf wins almost everywhere except that it has a higher barrier to entry for newbie bot developers. `ntba` is much easier to pick up.

[**grammY**](https://grammy.dev) is a bot framework that outperforms Telegraf in pretty much every respect.

The key difference is that it is substantially easier to learn, while being close enough to Telegraf that migrations from there are seamless. This is confirmed by many people who join the community chat and share their experiences.

Powerful: `grammY > Telegraf 4 > Telegraf 3 >> NTBA`
Easy to use: `grammY > Telegraf 3 > Telegraf 4` and also `NTBA > Telegraf 3 > Telegraf 4`

I assume that NTBA is easier to use than grammY for tiny bots, but virtually anything that goes beyond a simple echo bot is probably much simpler to implement with grammY. I'm interested in what you guys think.

Check out this comparison: <https://grammy.dev/resources/comparison.html>

**Reply by kamikazechaser** [reply]
Very cool indeed. Looks like the middleware stack (specifically commands) uses [linear routing (Array.some ?) with 1st match priority](https://grammy.dev/guide/middleware.html#the-middleware-stack). In the case of say 250 commands:

```js
bot.command("/start")
// 248 more commands
bot.command("/cmd249")
```

If a user invokes /cmd249, wouldn't that checking be inefficient?

**Reply by KnorpelSenf** [reply]
Yes, that is correct. This optimisation is yet to be made.

It would probably make a lot of sense to perform some transformations on the middleware tree to check filter queries more efficiently, too. This is non-trivial but possible.

I did some performance measurements with 1K checks and we're still talking about a few dozen microseconds per update for evaluating the middleware. That's why I prioritise other improvements for now. For example, smarter session management lets us drop a network call in many cases.

**Reply by kamikazechaser** [reply]
A nice. Because this is a legit use case. I have come across some crypto bots which have hundreds of commands. A trie would be great, I tried some experiment but wasn't really successful with it. Game changing library nonetheless.

**Reply by NormandoHall** [reply]
Well.... time to migrate to grammY!!!! Superb documentation!!!!! Thanks!!!

**Reply by KnorpelSenf** [reply]
@kamikazechaser in the meantime there was the 1.3 release of grammY which does some optimisations to [filter queries](https://grammy.dev/guide/filter-queries.html) (not commands). We now compile queries on startup to a callable trie function that performs checks with the minimal number of comparisons. As grammY has knowledge about the Bot API, it can make some smart assumptions so that it ends up performing slightly less checks than telegraf despite being significantly more powerful and expressive. I'd love to see a performance test of this. `bot.on()` listeners are likely to be used even more than commands. Here is the implementation if you're interested: https://github.com/grammyjs/grammY/blob/9aafd952a9272caae8de0d87ed8e993ce4bffd74/deno/src/filter.ts#L31
